[#339] Add pargraph about translation concatenation (#153)
infinum/frontend-handbook: Angular/Angular guidelines and best practices/Formatting, naming and best practices.md

></my-app-article-details>
```

## Avoid concatenation of translation keys

> Usual requirement for applications could be that the application needs to be localized. Regardless of the library which is used for the fetching of localized values,

Reviewer: ```suggestion
A common requirement for applications is that the application must to be localized. Regardless of the library which is used for fetching the translations,
```

> we should avoid concatenation of keys used for fetching translations. One of the issues is keeping the file with translations clean, it could prove very tiresome and hard even, to keep that file in sync with translations that are used throughout the application when you have to search for partial strings instead of searching for the whole string. One of the most tempting cases for concatenation is when you iterate over some enum and match those values with the ones received from the backend service. 

Reviewer: ```suggestion
we should avoid concatenation of translation keys. Concatenating translation keys makes managing of translation keys harder. It could prove very tiresome to keep the translations file in sync with translations that are used throughout the application, if you have to search for partial strings instead of searching for the whole string. One of the most tempting cases for using concatenation is when you iterate over some enum and match those values with the ones received from the backend service.
```

```ts
enum ApprovalStatus{
  Approved = 'Approved',
  Rejected = 'Rejected',
  NotApplicable = 'N/A',
}
```

```html
<mat-radio-button *ngFor="let approvalStatus of approvalStatusOptions">

>   {{'enums.approvalStatus.' + 'approvalStatus' | translate }}

Reviewer: ```suggestion
  {{ 'enums.approvalStatus.' + approvalStatus | translate }}
```

</mat-radio-button>
```

> In the above example you also have to make sure that you create "approvalStatusOptions" structure which will be iterable to be able to be used in "ngFor" directive.

Reviewer: ```suggestion
In the example above, you also have to make sure that you create `approvalStatusOptions` structure which will be iterable and used in the `ngFor` directive.
```

> In the cases where enum keys don't match their values there is also a question of how will you store translation keys in a way that will make the most sense. Should you use enum keys in the translations file, or should you use enum values corresponding to that key.

Reviewer: ```suggestion
In cases where enum keys don't match their values, there is also the question of storing translation keys in a way that will make the most sense. Should you use enum keys in the translations file, or should you use enum values corresponding to that key?
```

> Another case that emphasizes the issue with this approach is when you need to add some new keys, you have to verify that you've added them, in potentially multiple places and that everything will click together as intended.

Reviewer: ```suggestion
Another case that emphasizes the issue with this approach is when you need to add some new keys. You have to verify that you've added them to potentially multiple places and that everything will work together as intended.
```

> Approach i.e. pattern which proved to work for us is "Translatable enum" pattern which relies on using typescript Record utility type and for which we've implemented EnumPropertyPipe which is available in [ngx-nuts-and-bolts](https://infinum.github.io/ngx-nuts-and-bolts/docs/pipes/enum-property) library.

Reviewer: ```suggestion
A pattern that proved to work well for us is the "translatable enum" pattern which relies on using TypeScript's `Record` utility type and for which we've implemented `EnumPropertyPipe` that is available in the [ngx-nuts-and-bolts](https://infinum.github.io/ngx-nuts-and-bolts/docs/pipes/enum-property) library.
```

```ts

> enum ITranslatableEnum{

Reviewer: ```suggestion
enum ITranslatableEnum {
```

  translationKey: string;
}

export enum ApprovalStatus {
  Approved = 'Approved',
  Rejected = 'Rejected',
  NotApplicable = 'N/A',
}

export const approvalStatusData: Record<ApprovalStatus, ITranslatableEnum> = {
	[ApprovalStatus.Approved]: {
		translationKey: 'enums.approvalStatus.approved',
	},
	[ApprovalStatus.Rejected]: {
		translationKey: 'enums.approvalStatus.rejected',
	},
  [ApprovalStatus.NotApplicable]: {
		translationKey: 'enums.approvalStatus.notApplicable',
	},
};

@Component({
  selector: 'demo-component',

>   template: `<mat-radio-button *ngFor="let approvalStatus of approvalStatusOptions">{{ approvalStatus | enumProperty: approvalStatusData | translate}}`

Reviewer: ```suggestion
  template: `
  <mat-radio-button *ngFor="let approvalStatus of approvalStatusOptions">
    {{ approvalStatus | enumProperty: approvalStatusData | translate}}
  </mat-radio-button>
  `
```

})
class DemoComponent{

  public readonly ApprovalStatus = ApprovalStatus;
  public readonly approvalStatusData = approvalStatusData;
}
```

> By using this approach, your translation keys are all in one place which eases refactoring and keeping the translation file clean and up to date. 

Reviewer: ```suggestion
By using this approach, your translation keys are all in one place, easing refactoring and keeping the translation file clean and up to date. You will also get compilation errors if a new enum is added without adding a corresponding entry with the translation key in the record.

You can further expand on this pattern by adding additional "metadata" about the enum to the record. For example, a `sortingIndex` property that is taken into account when the enum values are rendered as dropdown options.
```
